### PR TITLE
[multistage] add rel trait rule for individual hep optimization

### PR DESCRIPTION
**Pull request by walterddr**
Allow propagation of partition distribution trait info across the tree to be used during Physical Planning phase.
This can be used to
- unnecessary shuffling can be changed into direct (if the list of servers are identical)
- multi-exchanges can be collapsed 
- allow more leaf-stage optimization rules to be run.

Follow up
===
- This PR only allows LEFT-side trait propagation for JOIN nodes; 
- This PR does not support trait propagation for SET-UNION / WINDOW nodes.

### CODE REVIEW

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotRelDistributionTraitRule.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+import org.apache.calcite.util.mapping.MappingType;
+import org.apache.calcite.util.mapping.Mappings;
+import org.apache.pinot.query.planner.plannode.AggregateNode;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+
+/**
+ * Special rule for Pinot, this rule is fixed to always insert exchange after JOIN node.
+ */
+public class PinotRelDistributionTraitRule extends RelOptRule {
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotRelDistributionTraitRule.java:62]: 1. this rule should ideally propagate all RelDistribution trait to all RelNodes.
2. when doing stage and worker/mailbox assignment, one can be sure to use the input of an LogicalExchange & the input to the LogicalExchange to determine whether the exchange can be a passthrough (fixing issues in https://github.com/apache/pinot/pull/11794)

CC @Jackie-Jiang @ankitsultana

**Comment by walterddr** [comment]
Note to self:
Exchange elimination cannot be done in this rule. it has to be done after the RelDistribution has propagated all the way through. then a top-down (instead of bottom-up) rule needs to apply to eliminate exchange in a 
- (RelNode -> Exchange -> RelNode) fashion for singleRel
- (RelNode -> left/right Exchange -> left/right RelNode) fashion for joins 

otherwise the exchange elimination is done with not enough info and can cause join to not function properly

**Comment by ankitsultana** [comment]
@walterddr : Is this ready for review?

**Comment by walterddr** [comment]
not yet @ankitsultana .we actually made some slight changes so that this PR doesn't depending on the other 2 PRs. marking it back to draft

**Comment by walterddr** [comment]
closing in favor of #11976

**Pull request closed without merging**